presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorNodePartitioningProvider.java

```diff
@@ -52,4 +56,61 @@ BucketFunction getBucketFunction(
             ConnectorPartitioningHandle partitioningHandle,
             List<Type> partitionChannelTypes,
             int bucketCount);
+
+    class ConnectorBucketNodeMap
+    {
+        private final int bucketCount;
+        private final Optional<Map<Integer, Node>> bucketToNode;
+
+        public static ConnectorBucketNodeMap createBucketNodeMap(int bucketCount)
+        {
+            return new ConnectorBucketNodeMap(bucketCount, Optional.empty());
+        }
+
+        public static ConnectorBucketNodeMap createBucketNodeMap(Map<Integer, Node> bucketToNode)
+        {
+            requireNonNull(bucketToNode, "bucketToNode is null");
+            if (bucketToNode.isEmpty()) {
+                throw new IllegalArgumentException("bucketToNode is empty");
+            }
+
+            return new ConnectorBucketNodeMap(
+                    bucketToNode.keySet().stream()
+                            .mapToInt(Integer::intValue)
+                            .max()
+                            .getAsInt() + 1,
+                    Optional.of(bucketToNode));
+        }
+
+        private ConnectorBucketNodeMap(int bucketCount, Optional<Map<Integer, Node>> bucketToNode)
+        {
+            if (bucketCount <= 0) {
+                throw new IllegalArgumentException("bucketCount must be positive");
+            }
+            if (bucketToNode.isPresent() && bucketToNode.get().size() != bucketCount) {
+                throw new IllegalArgumentException(format("Mismatched bucket count in bucketToNode (%s) and bucketCount (%s)", bucketToNode.get().size(), bucketCount));
+            }
+            this.bucketCount = bucketCount;
+            this.bucketToNode = requireNonNull(bucketToNode, "bucketToNode is null");
+        }
+
+        public int getBucketCount()
+        {
+            return bucketCount;
+        }
+
+        public boolean hasFixedMapping()
+        {
+            return bucketToNode.isPresent();
+        }
+
+        public Map<Integer, Node> getFixedMapping()
+        {
+            if (!bucketToNode.isPresent()) {
```

Comment:
Assuming you make a defensive copy above, the can be simplified to `return bucketToNode.orElseThrow(() -> new IllegalArgumentException("No fixed bucket to node mapping"));`